crates/sui-source-validation/src/lib.rs

```diff
@@ -396,3 +474,273 @@ fn local_modules(
+        );
+
+        let mut response = ureq::get(&url).call()?.into_reader();
+        let dest_tarball = dest_version.join(format!("{}.tgz", compiler_version));
+        debug!("tarball destination: {} ", dest_tarball.display());
+        if let Some(parent) = dest_tarball.parent() {
+            std::fs::create_dir_all(parent)
+                .map_err(|e| anyhow!("failed to create directory for tarball: {e}"))?;
+        }
+        let mut dest_file = File::create(&dest_tarball)?;
+        io::copy(&mut response, &mut dest_file)?;
+
+        // Extract the tarball using the tar crate
+        let tar_gz = File::open(&dest_tarball)?;
+        let tar = flate2::read::GzDecoder::new(tar_gz);
+        let mut archive = Archive::new(tar);
+        archive
+            .unpack(&dest_version)
+            .map_err(|e| anyhow!("failed to untar compiler binary: {e}"))?;
+
+        set_executable_permission(dest_binary_os)?;
+    }
+
+    debug!(
+        "sui move build --default-move-edition {} --default-move-flavor {} -p {}",
+        edition.to_string().as_str(),
+        flavor.to_string().as_str(),
+        root.display()
+    );
+    info!(
+        "{} {} (compiler @ {})",
+        "BUILDING".bold().green(),
+        dep_name.as_str(),
+        compiler_version.yellow()
+    );
+    Command::new(dest_binary_os)
+        .args([
+            OsStr::new("move"),
+            OsStr::new("build"),
+            OsStr::new("--default-move-edition"),
+            OsStr::new(edition.to_string().as_str()),
+            OsStr::new("--default-move-flavor"),
+            OsStr::new(flavor.to_string().as_str()),
+            OsStr::new("-p"),
+            OsStr::new(root.as_path()),
+        ])
+        .output()
+        .map_err(|e| {
+            anyhow!("failed to build package from compiler binary {compiler_version}: {e}",)
+        })?;
+    Ok(())
+}
+
+fn detect_platform() -> anyhow::Result<String> {
+    let s = match (std::env::consts::OS, std::env::consts::ARCH) {
+        ("macos", "aarch64") => "macos-arm64",
+        ("macos", "x86_64") => "macos-x86_64",
+        ("linux", "x86_64") => "ubuntu-x86_64",
+        ("windows", "x86_64") => "windows-x86_64",
+        (os, arch) => bail!("unsupported os {os} and arch {arch}"),
+    };
+    Ok(s.into())
+}
+
+#[cfg(unix)]
+fn set_executable_permission(path: &OsStr) -> anyhow::Result<()> {
+    use std::fs;
+    use std::os::unix::prelude::PermissionsExt;
+    let mut perms = fs::metadata(path)?.permissions();
+    perms.set_mode(0o755);
+    fs::set_permissions(path, perms)?;
+    Ok(())
+}
+
+#[cfg(not(unix))]
+fn set_executable_permission(path: &OsStr) -> anyhow::Result<()> {
+    Command::new("icacls")
+        .args([path, OsStr::new("/grant"), OsStr::new("Everyone:(RX)")])
+        .status()?;
+    Ok(())
+}
+
+fn decode_bytecode_file(
+    root_path: PathBuf,
+    package_name: &Symbol,
+    bytecode_path_str: &str,
+) -> anyhow::Result<CompiledUnitWithSource> {
```

Comment:
This is basically a free function of `OnDiskCompiledPackage`'s [decode_unit](https://sourcegraph.com/github.com/MystenLabs/sui@4256606ad498918c5a0f4594dd6b5fb9479464df/-/blob/external-crates/move/crates/move-package/src/compilation/compiled_package.rs?L198)